sigsci_site_manager/user.py

```diff
@@ -0,0 +1,166 @@
+import sys
+
+from sigsci_site_manager.api import init_api
+from sigsci_site_manager.util import add_new_user
+
+
+def do_list_users(args):
+    api = init_api(args.username, args.password, args.token, args.corp,
+                   args.dry_run)
+    if args.site_name:
+        api.site = args.site_name
+        users = api.get_site_members()
+    else:
+        users = api.get_corp_users()
+
+    cols = ['email', 'role', 'status', 'name']
+    colFormat = "%45s %10s %10s %30s"
+    if users:
+        line_entries = []
+        if args.site_name:
+            print('Site: %s' % args.site_name)
+            print(colFormat % (cols[0], 'site' + cols[1],
+                               cols[2], cols[3]))
+            for user in users['data']:
+                line_entries.append(colFormat % (user['user'][cols[0]], user[cols[1]],
+                                                 user['user'][cols[2]], user['user'][cols[3]]))
+            print_sorted_array(line_entries)
+        else:
+            print(colFormat % (cols[0], 'corp' + cols[1],
+                               cols[2], cols[3]))
+            for user in users['data']:
+                line_entries.append(colFormat % (user[cols[0]], user[cols[1]],
+                                                 user[cols[2]], user[cols[3]]))
+        print_sorted_array(line_entries)
+
+
+def do_list_membership(args):
+    api = init_api(args.username, args.password, args.token, args.corp,
+                   args.dry_run)
+    if args.email_id:
+        memberships = api.get_memberships(args.email_id)
+        cols = ['role', 'site']
+        colFormat = "%10s %s"
+        if memberships:
+            line_entries = []
+            print(colFormat % (cols[0], cols[1]))
+            for member in memberships['data']:
+                line_entries.append(colFormat % (member[cols[0]],
+                                                 "%s [%s]" % (member[cols[1]]['displayName'],
+                                                              member[cols[1]]['name'])))
+            print_sorted_array(line_entries)
+
+
+def do_add_user(args):
+    api = init_api(args.username, args.password, args.token, args.corp,
+                   args.dry_run)
+
+    input_file = resolve_input_file(args.file_name)
+    location_string = 'corp'
+    if args.site_name:
+        api.site = args.site_name
+        location_string = args.site_name
+
+    if args.email_id:
+        data = resolve_user_data_role(api, args.role_name)
+        if args.dry_run:
+            print("Adding %s to %s with %s role" %
+                  (args.email_id, location_string, data['role']))
+        add_new_user(api, args.email_id, args.role_name, args.api_user)
+    if input_file:
+        add_users_from_file(input_file, api, args, location_string)
```

Comment:
To reduce code duplication, the `add_users_from_file()` method should should read the file and return a list of users and roles (turn it into `get_users_from_file()`. If the `email_id` is specified instead create a list with a single user and role entry. Then, you can loop through that list and only need to call `add_new_user()` in one place.

Something like:
```python
user_list = []
if args.email_id:
    user_list = [(args.email_id, resolve_user_data_role(api, args.role_name)['role'])]
else if input_file:
    user_list = get_users_from_file(input_file, api, args)

for user in user_list:
    if args.dry_run:
        print("Adding %s to %s with %s role" %
                  (user[0], location_string, user[1]))
        add_new_user(api, user[0], args.role_name, user[1])
```

Reply:
addressed in comit 340cc86